src/libraries/Native/Unix/System.Native/pal_networking.c

```diff
@@ -432,12 +421,144 @@ int32_t SystemNative_GetHostEntryForName(const uint8_t* address, HostEntry* entr
     return ret;
 }
 
+#if HAVE_GETADDRINFO_A
+struct GetAddrInfoAsyncState
+{
+    struct gaicb gai_request;
+    struct gaicb* gai_requests;
+    struct sigevent sigevent;
+
+    uint8_t* address;
+    HostEntry* entry;
+    GetHostEntryForNameCallback callback;
+};
+
+static void GetHostEntryForNameAsyncComplete(sigval_t context)
+{
+    struct GetAddrInfoAsyncState* state = (struct GetAddrInfoAsyncState*)context.sival_ptr;
+
+    atomic_thread_fence(memory_order_acquire);
+
+    GetHostEntryForNameCallback callback = state->callback;
+
+    int ret = ConvertGetAddrInfoAndGetNameInfoErrorsToPal(gai_error(&state->gai_request));
+
+    if (ret == 0)
+    {
+        const uint8_t* address = state->address;
+        struct addrinfo* info = state->gai_request.ar_result;
+
+        ret = GetHostEntries(address, info, state->entry);
+    }
+
+    if (callback != NULL)
+    {
+        callback(state->entry, ret);
+    }
+
+    free(state->address);
+    free(state);
+}
+#endif
+
+int32_t SystemNative_PlatformSupportsGetAddrInfoAsync()
+{
+    return HAVE_GETADDRINFO_A;
+}
+
+// Get all address families and the canonical name
+static const struct addrinfo s_hostEntryForNameHints = {.ai_family = AF_UNSPEC, .ai_flags = AI_CANONNAME};
+
+int32_t SystemNative_GetHostEntryForName(const uint8_t* address, HostEntry* entry)
+{
+    if (address == NULL || entry == NULL)
+    {
+        return GetAddrInfoErrorFlags_EAI_BADARG;
+    }
+
+    struct addrinfo* info = NULL;
+
+    int result = getaddrinfo((const char*)address, NULL, &s_hostEntryForNameHints, &info);
+    if (result != 0)
+    {
+        return ConvertGetAddrInfoAndGetNameInfoErrorsToPal(result);
+    }
+
+    return GetHostEntries(address, info, entry);
+}
+
+int32_t SystemNative_GetHostEntryForNameAsync(const uint8_t* address, HostEntry* entry, GetHostEntryForNameCallback callback)
+{
+#if HAVE_GETADDRINFO_A
+    if (address == NULL)
+    {
+        return GetAddrInfoErrorFlags_EAI_BADARG;
+    }
+
+    struct GetAddrInfoAsyncState* state = malloc(sizeof(*state));
+
+    if (state == NULL)
+    {
+        return GetAddrInfoErrorFlags_EAI_MEMORY;
+    }
+
+    char* localAddress = strdup((const char*)address);
```

Comment:
malloc() result is NULL-checked but strdup() isn't

Reply:
You could do both allocations here in a single one by using a flexible member in struct GetAddrInfoAsyncState:

```c
struct GetAddrInfoAsyncState {
  ...
  char address[];
};

size_t addrlen = strlen(address);
struct GetAddrInfoAsyncState* state = malloc(sizeof(*state) + addrlen + 1);

memcpy(state->address, address, addlen+1);
```